### PR TITLE
expression: rewrite builtin function: ADDDATE, SUBDATE

### PR DESCRIPTION
**Pull request by zz-jason**
to #4080

### CODE REVIEW

**Comment by zz-jason** [comment]
@XuHuaiyu PTAL

File: expression/builtin_time.go [hunk 1]
```diff
@@ -1918,139 +1932,541 @@ func (b *builtinExtractDurationSig) evalInt(row []types.Datum) (int64, bool, err
...
+}
+type builtinDateAddDatetimeIntSig struct {
+	baseTimeBuiltinFunc
 }
 
-func (b *builtinDateArithSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
-	if err != nil {
-		return d, errors.Trace(err)
+func dateArithmeticalGetStringDate(ctx context.Context, args []Expression, row []types.Datum, unit string) (types.Time, bool, error) {
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: add comments for these functions.

File: expression/builtin_time.go [hunk 2]
```diff
@@ -1918,139 +1932,541 @@ func (b *builtinExtractDurationSig) evalInt(row []types.Datum) (int64, bool, err
...
 	}
-	nodeDate := args[0]
-	nodeIntervalValue := args[1]
-	nodeIntervalUnit := args[2].GetString()
-	if nodeIntervalValue.IsNull() {
-		return
+
+	date, err := types.ParseTime(dateStr, dateTp, types.MaxFsp)
+	if err != nil {
+		return types.Time{}, true, errors.Trace(handleInvalidTimeError(ctx, err))
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: we may `return date,  err != nil, errors.Trace(xxx)` without this check?

File: expression/builtin_time.go [hunk 3]
```diff
@@ -1918,139 +1932,541 @@ func (b *builtinExtractDurationSig) evalInt(row []types.Datum) (int64, bool, err
...
 	if err != nil {
-		return d, errInvalidOperation.Gen("DateArith invalid args, need date but get %T", nodeDate)
-	}
-	if value.IsNull() {
-		return d, errInvalidOperation.Gen("DateArith invalid args, need date but get %v", value.GetValue())
+		return types.Time{}, true, errors.Trace(handleInvalidTimeError(ctx, err))
 	}
-	if value.Kind() != types.KindMysqlTime {
-		return d, errInvalidOperation.Gen("DateArith need time type, but got %T", value.GetValue())
+	return date, false, nil
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: ditto

File: expression/builtin_time.go [hunk 4]
```diff
@@ -1918,139 +1932,541 @@ func (b *builtinExtractDurationSig) evalInt(row []types.Datum) (int64, bool, err
...
-		if nodeIntervalValue.Kind() == types.KindString {
-			interval = fmt.Sprintf("%v", nodeIntervalValue.GetString())
-		} else {
-			ii, err1 := nodeIntervalValue.ToInt64(sc)
-			if err1 != nil {
-				return d, errors.Trace(err1)
-			}
-			interval = fmt.Sprintf("%v", ii)
-		}
+		date.Type = mysql.TypeDate
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: move this to line 2051, no need to use `else`

File: expression/builtin_time.go [hunk 5]
```diff
@@ -1918,139 +1932,541 @@ func (b *builtinExtractDurationSig) evalInt(row []types.Datum) (int64, bool, err
 	return res, false, errors.Trace(err)
 }
 
-type dateArithFunctionClass struct {
-	baseFunctionClass
+var reg = regexp.MustCompile(`[\d]+`)
 
-	op ast.DateArithType
+func parseInterval4AddSubDate4DAY(interval string) string {
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: add a comment for this func.

File: expression/builtin_time.go (hunk 1 above)
Comment (XuHuaiyu) [inline expression/builtin_time.go]: extract a base class to implement these functions for limit the usage of them.

File: expression/builtin_time.go [hunk 6]
```diff
@@ -1915,142 +1929,600 @@ func (b *builtinExtractDurationSig) evalInt(row []types.Datum) (int64, bool, err
...
+	if isNull || err != nil {
+		return types.Time{}, true, errors.Trace(err)
+	}
+
+	date, isNull, err := b.getDateFromDatetime(b.ctx, b.args, row, unit)
+	if isNull || err != nil {
+		return types.Time{}, true, errors.Trace(err)
+	}
+
+	interval, isNull, err := b.getIntervalFromInt(b.ctx, b.args, row, unit)
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: getIntervalFromString or getIntervalFromInt?

File: expression/builtin_time.go [hunk 7]
```diff
@@ -1915,142 +1929,600 @@ func (b *builtinExtractDurationSig) evalInt(row []types.Datum) (int64, bool, err
 		return 0, isNull, errors.Trace(err)
 	}
 	res, err := types.ExtractDurationNum(&dur, unit)
-	return res, false, errors.Trace(err)
+	return res, err != nil, errors.Trace(err)
 }
 
-type dateArithFunctionClass struct {
-	baseFunctionClass
+type baseDateArithmitical struct {
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: add a comment for this class and the attribute.

File: expression/builtin_time.go [hunk 8]
```diff
@@ -1915,142 +1929,600 @@ func (b *builtinExtractDurationSig) evalInt(row []types.Datum) (int64, bool, err
...
+	}
+	return interval, false, nil
+}
+
+func (du *baseDateArithmitical) getIntervalFromInt(ctx context.Context, args []Expression, row []types.Datum, unit string) (string, bool, error) {
+	interval, isNull, err := args[1].EvalInt(row, ctx.GetSessionVars().StmtCtx)
+	if isNull || err != nil {
+		return "", true, errors.Trace(err)
+	}
+	return fmt.Sprintf("%v", interval), false, nil
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]:
``` go
func BenchmarkGenStrSprintf(b *testing.B) {
	for i := 0; i < b.N; i++ {
		fmt.Sprintf("%v", 12345)
	}
}

func BenchmarkGenStrStrconv(b *testing.B) {
	for i := 0; i < b.N; i++ {
		strconv.FormatInt(12345, 10)
	}
}
```

BenchmarkGenStrSprintf-4       10000000               160 ns/op
BenchmarkGenStrStrconv-4        30000000                47.3 ns/op

File: expression/integration_test.go [hunk 9]
```diff
@@ -1518,6 +1518,79 @@ func (s *testIntegrationSuite) TestTimeBuiltin(c *C) {
...
 
+	// for adddate, subdate
+	dateArithmeticalTests := []struct {
+		Date      string
+		Interval  string
+		Unit      string
+		AddResult string
+		SubResult string
+	}{
+		{"\"2011-11-11\"", "1", "DAY", "2011-11-12", "2011-11-10"},
```
Comment (XuHuaiyu) [inline expression/integration_test.go]: why escape \" ?

Comment (zz-jason) [reply expression/integration_test.go]: means it is a string literal

File: expression/builtin_time.go [hunk 10]
```diff
@@ -2487,7 +2490,7 @@ func (b *builtinSubDateDatetimeStringSig) evalTime(row []types.Datum) (types.Tim
 		return types.Time{}, true, errors.Trace(err)
 	}
 
-	interval, isNull, err := b.getIntervalFromInt(b.ctx, b.args, row, unit)
+	interval, isNull, err := b.getIntervalFromString(b.ctx, b.args, row, unit)
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: add a test case to cover this

**Comment by XuHuaiyu** [comment]
PTAL @winkyao

**Comment by XuHuaiyu** [comment]
LGTM

File: expression/builtin_time.go [hunk 11]
```diff
@@ -1918,139 +1932,541 @@ func (b *builtinExtractDurationSig) evalInt(row []types.Datum) (int64, bool, err
...
+	if !types.IsDateFormat(dateStr) || types.IsClockUnit(unit) {
+		dateTp = mysql.TypeDatetime
 	}
-	nodeDate := args[0]
-	nodeIntervalValue := args[1]
-	nodeIntervalUnit := args[2].GetString()
-	if nodeIntervalValue.IsNull() {
-		return
+
+	date, err := types.ParseTime(dateStr, dateTp, types.MaxFsp)
```
Comment (winkyao) [inline expression/builtin_time.go]: types.MaxFsp? why not use dateStr decimal length?

File: expression/builtin_time.go [hunk 12]
```diff
@@ -1918,139 +1932,541 @@ func (b *builtinExtractDurationSig) evalInt(row []types.Datum) (int64, bool, err
...
+}
+
+func dateArithmeticalGetIntDate(ctx context.Context, args []Expression, row []types.Datum, unit string) (types.Time, bool, error) {
+	dateInt, isNull, err := args[0].EvalInt(row, ctx.GetSessionVars().StmtCtx)
+	if isNull || err != nil {
+		return types.Time{}, true, errors.Trace(err)
+	}
+
+	dateTp := mysql.TypeDate
+	if t, err := types.ParseTimeFromInt64(dateInt); err == nil {
```
Comment (winkyao) [inline expression/builtin_time.go]: types.ParseTimeFromInt64 is for what? if err == nil, why not return t directlly?

File: expression/builtin_time.go [hunk 13]
```diff
@@ -1918,139 +1932,541 @@ func (b *builtinExtractDurationSig) evalInt(row []types.Datum) (int64, bool, err
...
-	year, month, day, dur, err := types.ExtractTimeValue(nodeIntervalUnit, interval)
+	return interval, false, nil
+}
+
+func dateArithmeticalGetIntInterval(ctx context.Context, args []Expression, row []types.Datum, unit string) (string, bool, error) {
+	interval, isNull, err := args[1].EvalInt(row, ctx.GetSessionVars().StmtCtx)
+	if isNull || err != nil {
+		return "", true, errors.Trace(err)
+	}
+	return fmt.Sprintf("%v", interval), false, nil
```
Comment (winkyao) [inline expression/builtin_time.go]: use strconv.Itoa?

File: expression/builtin_time.go [hunk 14]
```diff
@@ -1915,142 +1929,606 @@ func (b *builtinExtractDurationSig) evalInt(row []types.Datum) (int64, bool, err
...
 	}
-	sig := &builtinDateArithSig{newBaseBuiltinFunc(args, ctx), c.op}
-	return sig.setSelf(sig), nil
+
+	dateTp := mysql.TypeDate
+	if !types.IsDateFormat(dateStr) || types.IsClockUnit(unit) {
+		dateTp = mysql.TypeDatetime
+	}
+
+	date, err := types.ParseTime(dateStr, dateTp, types.MaxFsp)
```
Comment (winkyao) [inline expression/builtin_time.go]: types.MaxFsp? why not use dateStr decimal length?

File: expression/builtin_time.go [hunk 15]
```diff
@@ -168,11 +170,23 @@ var (
...
+	_ builtinFunc = &builtinSubDateIntStringSig{}
+	_ builtinFunc = &builtinSubDateIntIntSig{}
+	_ builtinFunc = &builtinSubDateDatetimeStringSig{}
+	_ builtinFunc = &builtinSubDateDatetimeIntSig{}
 )
 
 // handleInvalidTimeError reports error or warning depend on the context.
 func handleInvalidTimeError(ctx context.Context, err error) error {
-	if err == nil || !terror.ErrorEqual(err, types.ErrInvalidTimeFormat) {
+	if err == nil || !(terror.ErrorEqual(err, types.ErrInvalidTimeFormat) || types.ErrIncorrectDatetimeValue.Equal(err)) {
```
Comment (winkyao) [inline expression/builtin_time.go]: types. ErrInvalidTimeFormat.Equal(err) ?

Comment (zz-jason) [reply expression/builtin_time.go]: `ErrInvalidTimeFormat` do not have a method named `Equal`

File: expression/builtin_time.go (hunk 14 above)
Comment (zz-jason) [reply expression/builtin_time.go]: dateStr's Decimal will always be -1

**Pull request merged**